cluster-test/test-orchestrator/test_yaml_parser.go

```diff
@@ -0,0 +1,308 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"gopkg.in/yaml.v2"
+)
+
+// testYaml is used to unmarshal test declared in the yaml files.
+type testYaml struct {
+	Config    configYaml
+	Scenarios []*scenarioYaml
+}
+
+type configYaml struct {
+	// TODO(wieger): define
+}
+
+// scenarioYaml captures the information of a scenario.
+type scenarioYaml struct {
+	Name string
+	Size string
+	Desc string
+
+	Script  []map[string]string
+	Measure []string
+	Runs    [][]string
+}
+
+func parse(bts []byte) (scns []*Scenario, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			scns = nil
+			err = errors.New(fmt.Sprint(r))
+		}
+	}()
+
+	return parseScenarios(bts), nil
+}
+
+func parseScenarios(bts []byte) []*Scenario {
+	testYaml := &testYaml{}
+	err := yaml.Unmarshal([]byte(bts), testYaml)
+	if err != nil {
+		panic("failed to unmarshal scenario yaml")
+	}
+
+	return extractScenarios(testYaml)
+}
+
+// extractScenarios returns a scenario for every element in the runs list.
+func extractScenarios(runs *testYaml) []*Scenario {
+	var result []*Scenario
+	for _, scenarioData := range runs.Scenarios {
+		for i := 1; i < len(scenarioData.Runs); i++ {
+			s := extractScenario(scenarioData, i)
+			result = append(result, s)
+		}
+	}
+
+	return result
+}
+
+// extractScenario returns a scenario given the index of a specific run.
+func extractScenario(data *scenarioYaml, runIx int) *Scenario {
+	varsData := data.Runs[0]
+	runData := data.Runs[runIx]
+	defer wrapPanicf("Failed to parse scenario '%s'", data.Name)
+	defer wrapPanicf("in run %d, [%v] = [%v]", runIx, strings.Join(varsData, ", "), strings.Join(runData, ", "))
+
+	if len(varsData) != len(runData) {
+		msg := fmt.Sprintf("var count of run %v should match var count of %v", runData, varsData)
+		panic(msg)
+	}
+
+	// don't find and replace on name
+	name := data.Name
+	desc := replace(data.Desc, varsData, runData)
+	sizeStr := replace(data.Size, varsData, runData)
+
+	// extract size
+	size, err := strconv.Atoi(sizeStr)
+	if err != nil {
+		panic("size convert: " + err.Error())
+	}
+
+	// extract script
+	labels, cmds := extractScript(data.Script, varsData, runData)
+	script := parseScript(labels, cmds)
```

Comment:
Because the functions where bigger previously it got a bit unwieldy having them in one function. I think it still makes sense to have them split, maybe not as much as previously though.